docs/design/byos.md

```diff
@@ -1,13 +1,91 @@
+
+The theme throughout these scenarios is that the customer wants hosted elasticity but customization beyond what Hosted offers.
+
+### Custom image
+
+1. Customer wants Windows 2016 with VS 2019 Preview. (We only offer certain combos like Win2016 + VS2017 and Win2019 + VS2019.)
+2. Customer wants to pin a specific set of tools and dependencies, preconfigured on the image.
+3. Customer wants to bring their own pre-configured image.
+
+### Custom SKU
+
+1. Customer wants more memory, more processor, or more IO than our native images.
+2. [Customer](https://github.com/MicrosoftDocs/vsts-docs/issues/2985) wants an NCv2 VM with particular instruction sets for machine learning. (It's niche enough that we won't stand up dedicated hosted pools, but broad enough to be very interesting for our business.)
+3. Customer wants additional storage attached to the VM. *(Real scenario from a medium-sized customer)*
+4. Customer wants their own network topology or Active Directory configuration.
+
+### Custom elasticity
+
+1. Customer wants to run several consecutive jobs on an agent to take advantage of things like incremental source and machine-level package caches. But, they don't want to run unnecessary VMs overnight when there's no load. They want to specify minimum and maximum # of agents associated with time ranges.
+2. Customer wants to run additional configuration or cache warmup before an agent beings accepting jobs. As additional agents are spun up, the customer has an opportunity to run a prep script that doesn't impact "pipeline runtime".
+
+## Alternatives considered
+
+### Custom / configurable SKU
+
+Offering additional Azure SKUs in Hosted will meet some of the above scenarios.
+Specifically, if all you need are additional resources (memory, compute, etc.), then a custom SKU works.
+
+Pros:
+- Easiest for the customer to understand
+- Potential business model around premium SKUs
+
+Cons:
+- Capacity planning and buildout becomes a lot more complex
+- If a customer wants a SKU we don't offer, they're out of luck until we add it
+- Does not offer custom elasticity, network environment, or image
+
+### Bring your own cloud (pool providers)
+
+Putting the entire agent alloc/release cycle in the hands of the customer offers them ultimate flexibility.
+
+Pros:
+- Maximum flexibility
+- Customer can select other infrastructure: on-prem, in a different hosting provider, etc.
+
+Cons:
+- Much heavier burden to understand, set up, and maintain
+- Customer has to swallow full complexity load to get any flexibility
+
+## "Clickstops" along the way
+
+Of course, we won't want to do a huge multi-month effort and then dump it on the community at once.
+We can deliver several intermediate levels of value on the way to a full solution.
+At every point, we can re-evaluate if the decisions we're making match customer needs at the time.
+
+**Stage 0**. Using a customer's Azure subscription, run pipelines against our native images using Hosted-like "throwaway" VMs.
```

Comment:
I really think for the vast majority of customers this is not what they want.  They want to have agents that exist for a longer period of time and benefit from the caching of packages and container layers locally.  To me the hosted throw away agents are the secondary scenario.

Reply:
Agree. That's why it's stage 0 / less than MVP. My thinking was to mitigate risk on our side. Not worth it?

Reply:
We absolutely want throw away agents rather than long lived, we'd just like them connected to our own vnets so they can manage other resources in the vnet without needing public IPs on everything.

Reply:
I agree with @hbuckle we only use hosted agents and a large number of them

We have to write custom scripts for temporary firewall rules at the moment which are problematic when it comes to clean-up

If we could simply attach them to our vnets that would be ideal

Reply:
Thanks @hbuckle and @steve-hawkins.

Reply:
@vtbassmatt I am not sure how it mitigates our risk in anyway.  In fact given some of the challenges we have seen with getting Azure Resources created and cleaned up I think this makes our risk much higher.

Reply:
@steve-hawkins I think both options are valid.  however, many of the customers I have spoken to have very large builds with many gigabytes of packages.  In addition to the vnet they want the benifit of improved performance of their builds by having the repo and packages on the machine.

Reply:
If it's done right, it's not one or the other.  It's a config setting on how long to live - run once, recycle nightly, etc...

Reply:
@chrisrpatterson if there's a thing we do that's known to be hard, shouldn't we isolate that hard thing and make sure we nail it? I might be missing something - being confident in our ability to manage VMs in foreign subscriptions seems fundamental to everything else.

@bryanmacfarlane see the later clickstops, that's definitely part of the completed feature.